homeassistant/components/kira.py

```diff
@@ -0,0 +1,137 @@
+
+from homeassistant.helpers import discovery
+import homeassistant.helpers.config_validation as cv
+
+from homeassistant.const import (
+    CONF_DEVICE,
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PORT,
+    CONF_SENSORS,
+    CONF_TYPE,
+    EVENT_HOMEASSISTANT_STOP,
+    STATE_UNKNOWN)
+
+
+REQUIREMENTS = ["pykira==0.1.0"]
+
+DOMAIN = 'kira'
+
+RECEIVERS = {}
+MODULES = {}
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_HOST = "0.0.0.0"
+DEFAULT_PORT = 65432
+
+CONF_CODE = "code"
+CONF_REPEAT = "repeat"
+CONF_REMOTES = "remotes"
+CONF_SENSOR = "sensor"
+CONF_REMOTE = "remote"
+
+CODES_YAML = '{}_codes.yaml'.format(DOMAIN)
+
+CODE_SCHEMA = vol.Schema({
+    vol.Required(CONF_NAME): cv.string,
+    vol.Required(CONF_CODE): cv.string,
+    vol.Optional(CONF_TYPE): cv.string,
+    vol.Optional(CONF_DEVICE): cv.string,
+    vol.Optional(CONF_REPEAT): cv.positive_int,
+})
+
+SENSOR_SCHEMA = vol.Schema({
+    vol.Optional(CONF_NAME, default=DOMAIN):
+        vol.Exclusive(cv.string, "sensors"),
+    vol.Optional(CONF_HOST, default=DEFAULT_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+})
+
+REMOTE_SCHEMA = vol.Schema({
+    vol.Optional(CONF_NAME, default=DOMAIN):
+        vol.Exclusive(cv.string, "remotes"),
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_SENSORS): [SENSOR_SCHEMA],
+        vol.Optional(CONF_REMOTES): [REMOTE_SCHEMA]})
+}, extra=vol.ALLOW_EXTRA)
+
+
+def load_codes(path):
+    codes = []
+    if os.path.exists(path):
+        with open(path) as f:
+            data = yaml.load(f)
+        for code in data:
+            try:
+                codes.append(CODE_SCHEMA(code))
+            except VoluptuousError as exception:
+                # keep going
+                _LOGGER.warning('Kira Code Invalid Data: %s', exception)
+    else:
+        with open(path, 'w') as f:
+            f.write('')
+    return codes
+
+
+def setup(hass, config):
+    """Setup KIRA capability."""
+    import pykira
+
+    sensors = config[DOMAIN].get(CONF_SENSORS, [])
+    remotes = config[DOMAIN].get(CONF_REMOTES, [])
+    if len(sensors) == 0 and len(remotes) == 0:
+        sensors.append({})
+    codes = load_codes(hass.config.path(CODES_YAML))
+
+    global RECEVIERS
+    global MODULES
+
+    for idx, moduleConf in enumerate(itertools.chain(sensors, remotes)):
+        # note: this is not the HA device name. it's just a unique name
+        # to ensure the component and platform can share information
+        moduleName = idx and ("%s_%d" % (DOMAIN, idx)) or DOMAIN
+        port = moduleConf.get(CONF_PORT, DEFAULT_PORT)
+        host = moduleConf.get(CONF_HOST, DEFAULT_HOST)
+
+        if moduleConf in sensors:
+            module = pykira.KiraReceiver(host, port)
+            RECEIVERS[moduleName] = module
+            module.start()
```

Comment:
This will, however the large majority of installations will require only one sensor as all Kira modules can usually send to the same UDP port. The only time a user would ever need to configure more than one is to distinguish between the same IR code sent to different Kira Modules.